web/src/components/storage/DevicesManager.js

```diff
@@ -0,0 +1,207 @@
+   * System device with the given SID.
+   *
+   * @param {Number} sid
+   * @returns {StorageDevice|undefined}
+   */
+  systemDevice(sid) {
+    return this.#device(sid, this.system);
+  }
+
+  /**
+   * Staging device with the given SID.
+   *
+   * @param {Number} sid
+   * @returns {StorageDevice|undefined}
+   */
+  stagingDevice(sid) {
+    return this.#device(sid, this.staging);
+  }
+
+  /**
+   * Whether the given device exists in system.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  existInSystem(device) {
+    return this.#exist(device, this.system);
+  }
+
+  /**
+   * Whether the given device exists in staging.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  existInStaging(device) {
+    return this.#exist(device, this.staging);
+  }
+
+  /**
+   * Whether the given device is going to be formatted.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  hasNewFilesystem(device) {
+    if (!device.filesystem) return false;
+
+    const systemDevice = this.systemDevice(device.sid);
+    const systemFilesystemSID = systemDevice?.filesystem?.sid;
+
+    return device.filesystem.sid !== systemFilesystemSID;
+  }
+
+  /**
+   * Whether the given device is going to be shrunk.
+   *
+   * @param {StorageDevice} device
+   * @returns {Boolean}
+   */
+  isShrunk(device) {
+    return this.shrinkSize(device) > 0;
+  }
+
+  /**
+   * Amount of bytes the given device is going to be shrunk.
+   *
+   * @param {StorageDevice} device
+   * @returns {Number}
+   */
+  shrinkSize(device) {
+    const systemDevice = this.systemDevice(device.sid);
+    const stagingDevice = this.stagingDevice(device.sid);
+
+    if (!systemDevice || !stagingDevice) return 0;
+
+    const amount = systemDevice.size - stagingDevice.size;
+    return amount > 0 ? amount : 0;
+  }
+
+  /**
+   * Disk devices and LVM volume groups used for the installation.
+   *
+   * @note The used devices are extracted from the actions.
+   *
+   * @returns {StorageDevice[]}
+   */
+  usedDevices() {
+    const isTarget = (device) => device.isDrive || device.type === "lvmVg";
+
+    // Check in system devices to detect removals.
+    const targetSystem = this.system.filter(isTarget);
+    const targetStaging = this.staging.filter(isTarget);
+
+    const sids = targetSystem.concat(targetStaging)
+      .filter(d => this.#isUsed(d))
+      .map(d => d.sid);
+
+    return compact(uniq(sids).map(sid => this.stagingDevice(sid)));
+  }
+
+  /**
+   * Devices deleted.
+   *
+   * @note The devices are extracted from the actions.
+   *
+   * @returns {StorageDevice[]}
+   */
+  deletedDevices() {
+    const sids = this.actions
+      .filter(a => a.delete)
+      .map(a => a.device);
+    const devices = sids.map(sid => this.systemDevice(sid));
+    return compact(devices);
+  }
+
+  /**
+   * Systems deleted.
+   *
+   * @returns {string[]}
+   */
+  deletedSystems() {
+    const systems = this.deletedDevices()
```

Comment:
This can add the same system twice if there is an action on the partition and on the disk (eg. deleting the partition table). Since both devices reference the same system.